src/Network/Wai/Auth/Executable.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+module Network.Wai.Auth.Executable
+  ( mkMain
+  , readAuthConfig
+  , parseProviders
+  , Port
+  ) where
+
+
+import           Data.Aeson.Types                     (parseEither)
+import qualified Data.HashMap.Strict                  as HM
+
+import           Data.String                          (fromString)
+import qualified Data.Text                            as T
+import           Data.Text.Encoding                   (encodeUtf8)
+import           Data.Yaml.Config                     (loadYamlSettings, useEnv)
+import           Network.HTTP.Client                  (Manager, newManager)
+import           Network.HTTP.Client.TLS              (tlsManagerSettings)
+import           Network.HTTP.ReverseProxy            (ProxyDest (..), WaiProxyResponse (WPRProxyDest),
+                                                       defaultOnExc, waiProxyTo)
+import           Network.Wai                          (Application)
+import           Network.Wai.Application.Static       (defaultFileServerSettings,
+                                                       ssAddTrailingSlash,
+                                                       ssRedirectToIndex,
+                                                       staticApp)
+import           Network.Wai.Auth.Config
+import           Network.Wai.Middleware.Auth
+
+import           Network.Wai.Middleware.Auth.Provider
+import           Network.Wai.Middleware.Redirect2tls
+
+import           Web.ClientSession                    (getKey)
+
+type Port = Int
+
+
+serviceToApp :: Manager -> Service -> IO Application
+serviceToApp _ (ServiceFiles FileServer {..}) =
+    return $ staticApp (defaultFileServerSettings $ fromString fsRootFolder)
+        { ssRedirectToIndex = fsRedirectToIndex
+        , ssAddTrailingSlash = fsAddTrailingSlash
+        }
+serviceToApp manager (ServiceProxy (ReverseProxy host port)) =
+    return $ waiProxyTo
+        (const $ return $ WPRProxyDest $ ProxyDest (encodeUtf8 host) port)
+        defaultOnExc
+        manager
+
+
+readAuthConfig :: FilePath -> IO AuthConfig
+readAuthConfig confFile = loadYamlSettings [confFile] [] useEnv
+
+
+parseProviders :: AuthConfig -> [ProviderParser] -> Providers
```

Comment:
This probably will be only called early on in `main` and executable should die at that point if there is really an error. But having `MonadThrow` instead of `error` will only make this function better, so I'll make that change.